core/client/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -437,50 +408,123 @@ private void logCacheStreamIOException(IOException e) {
   }
 
   /**
-   * Similar to {@link #checkAndAdvanceBlockInStream()}, but a specific position can be specified
-   * and the stream pointer will be at that offset after this method completes.
+   * Only updates {@link #mCurrentCacheStream}, {@link #mCurrentBlockInStream} and
+   * {@link #mStreamBlockId} to be in-sync the current block (i.e. {@link #getCurrentBlockId()}).
+   * This function can be called multiple times without side effect. It is recommended to be
+   * invoked before every read and seek.
    *
-   * @param newPos the new position to set the stream to
-   * @throws IOException if the stream at the specified position cannot be opened
+   * @throws IOException if the next cache stream or block stream cannot be created
    */
-  private void seekBlockInStream(long newPos) throws IOException {
-    long oldBlockId = getCurrentBlockId();
-    mPos = newPos;
-    closeCacheStream();
+  private void updateStreams() throws IOException {
     long currentBlockId = getCurrentBlockId();
-
-    if (oldBlockId != currentBlockId) {
+    if (shouldUpdateStreams(currentBlockId)) {
+      // The following two function handle negative currentBlockId (i.e. the end of file)
+      // correctly.
       updateBlockInStream(currentBlockId);
-      // Reading next block entirely.
-      if (mPos % mBlockSize == 0 && mShouldCacheCurrentBlock) {
-        try {
-          WorkerNetAddress address = mLocationPolicy.getWorkerForNextBlock(
-              mContext.getAlluxioBlockStore().getWorkerInfoList(), getBlockSizeAllocation(mPos));
-          mCurrentCacheStream = createCacheStream(currentBlockId, getBlockSize(mPos), address);
-        } catch (IOException e) {
-          logCacheStreamIOException(e);
-          mShouldCacheCurrentBlock = false;
-        } catch (AlluxioException e) {
-          LOG.warn(BLOCK_ID_NOT_CACHED, currentBlockId, e);
-          mShouldCacheCurrentBlock = false;
+      updateCacheStream(currentBlockId);
+      mStreamBlockId = currentBlockId;
+    }
+  }
+
+  /**
+   * Check whether block instream and cache outstream should be updated.
+   * This function is only called by {@link #updateStreams()}.
+   *
+   * @param currentBlockId cached result of {@link #getCurrentBlockId()}
+   * @return true if the block stream should be updated
+   */
+  private boolean shouldUpdateStreams(long currentBlockId) {
+    if (mCurrentBlockInStream == null || currentBlockId != mStreamBlockId) {
+      return true;
+    }
+    /*
+    if (mCurrentCacheStream != null) {
+      Preconditions.checkState(mCurrentBlockInStream.remaining() == mCurrentCacheStream.remaining(),
+          "BlockInStream and CacheStream is not sync-ed %d %d", mCurrentBlockInStream.remaining(),
+          mCurrentCacheStream.remaining());
+    }
```

Comment:
I commented this check because this won't work with UnknownLengthFileInStream. I need to talk with Gene about this.